fuel/transformers/image.py

```diff
@@ -0,0 +1,227 @@
+class MinimumImageDimensions(SourcewiseTransformer, ExpectsAxisLabels):
+    """Resize (lists of) images to minimum dimensions.
+
+    Parameters
+    ----------
+    data_stream : instance of :class:`AbstractDataStream`
+        The data stream to wrap.
+    minimum_shape : 2-tuple
+        The minimum `(height, width)` dimensions every image must have.
+        Images whose height and width are larger than these dimensions
+        are passed through as-is.
+    resample : str, optional
+        Resampling filter for PIL to use to upsample any images requiring
+        it. Options include 'nearest' (default), 'bilinear', and 'bicubic'.
+        See the PIL documentation for more detailed information.
+
+    Notes
+    -----
+    This transformer expects stream sources returning individual images,
+    represented as 2- or 3-dimensional arrays, or lists of the same.
+    The format of the stream is unaltered.
+
+    """
+    def __init__(self, data_stream, minimum_shape, resample='nearest',
+                 **kwargs):
+        self.minimum_shape = minimum_shape
+        try:
+            self.resample = getattr(Image, resample.upper())
+        except AttributeError:
+            raise ValueError("unknown resampling filter '{}'".format(resample))
+        super(MinimumImageDimensions, self).__init__(data_stream, **kwargs)
+
+    def transform_source_batch(self, batch):
+        self.verify_axis_labels(('batch', 'channel', 'height', 'width'),
+                                self.data_stream.axis_labels)
+        return [self.transform_source_example(im) for im in batch]
+
+    def transform_source_example(self, example):
+        self.verify_axis_labels(('channel', 'height', 'width'),
+                                self.data_stream.axis_labels)
+        if example.ndim > 3 or example.ndim < 2:
+            raise NotImplementedError
+        min_height, min_width = self.minimum_shape
+        original_height, original_width = example.shape[-2:]
+        if original_height < min_height or original_width < min_width:
+            dt = example.dtype
+            # If we're dealing with a colour image, swap around the axes
+            # to be in the format that PIL needs.
+            if example.ndim == 3:
+                im = example.tranpose(2, 0, 1)
+            else:
+                im = example
+            im = Image.fromarray(im)
+            width, height = im.size
+            # Based on the current width and height, generate
+            # dimensions that respect the minimum constraints but also
+            # match the old aspect ratio.
+            aspect = float(width) / height
+            if width < min_width:
+                width = min_width
+                height = int(width / aspect)
+            # N.B.: not elif. Aspect-ratio corrected height may be less
+            # than minimum requested height.
+            if height < min_height:
+                height = min_height
+                width = int(height * aspect)
```

Comment:
There's a more straightforward way to implement that logic: multiply `height` and `width` by `max(1, float(min_width) / width, float(min_height) / height)`. Here's the rationalization:

_Let **w** and **h** be the image width and height, and let **a** and **b** be the minimum width and height. We want to multiply **w** and **h** by the smallest number **r >= 1** such that the constraints **w >= a** and **h >= b** are satisfied._

_If both constraints are satisfied, then **r = 1** since nothing needs to be done. In addition, **w >= a** implies **a / w <= 1** and **h >= b** implies **b / h <= 1**, which means **r = max(1, a / w, b / h)**._

_If **w < a** but **h >= b**, then we need to scale up **w**. The minimum we can do is to bring it up to **a**, which means that **r = a / w**. In addition, **w < a** implies **a / w > 1** and **h >= b** implies **b / h <= 1**, which means **r = max(1, a / w, b / h)**. An analogous reasoning can be used to show that if only the **h** constraint is violated, we still have **r = max(1, a / w, b / h)**._

_Finally, if **w < a** and **h < b**, then we need to scale up both **w** and **h**. The minimum amount of scaling required for the two constraints is **r = a / w** and **r = b / h**, respectively. To satisfy both constraints, we need to choose the biggest scaling factor of those two. In other words, **r = max(a / w, b / h)**. In addition, **w < a** implies **a / w > 1** and **h < b** implies **b / h > 1**, which means **r = max(1, a / w, b / h)**._

_We conclude that **r = max(1, a / w, b / h)**._